spec/util/object_storage/miq_ftp_storage_spec.rb

```diff
@@ -0,0 +1,115 @@
+        end
+
+        expected_splitfiles.each do |filename|
+          expect(filename).to exist_on_ftp_server
+          expect(filename).to have_size_on_ftp_server_of(2.megabytes)
+        end
+      end
+
+      it "can take input from a command" do
+        expected_splitfiles = (1..5).map do |suffix|
+          "#{dest_path}.0000#{suffix}"
+        end
+
+        subject.add(dest_path.to_s, "2M") do |input_writer|
+          `#{Gem.ruby} -e "File.write('#{input_writer}', '0' * #{tmpfile_size})"`
+        end
+
+        expected_splitfiles.each do |filename|
+          expect(filename).to exist_on_ftp_server
+          expect(filename).to have_size_on_ftp_server_of(2.megabytes)
+        end
+      end
+
+      context "with slightly a slightly smaller input file than 10MB" do
+        let(:tmpfile_size) { 10.megabytes - 1.kilobyte }
+
+        it "properly chunks the file" do
+          expected_splitfiles = (1..10).map do |suffix|
+            "#{dest_path}.%<suffix>05d" % {:suffix => suffix}
+          end
+
+          # using pathnames this time
+          subject.add(source_path, dest_path.to_s, 1.megabyte)
+
+          expected_splitfiles[0, 9].each do |filename|
+            expect(filename).to exist_on_ftp_server
+            expect(filename).to have_size_on_ftp_server_of(1.megabytes)
+          end
+
+          last_split = expected_splitfiles.last
+          expect(last_split).to exist_on_ftp_server
+          expect(last_split).to have_size_on_ftp_server_of(1.megabyte - 1.kilobyte)
+        end
+      end
+    end
+
+    context "using a 'relative path'" do
+      include_examples "adding files", "path/to/file"
+    end
+
+    context "using a 'absolute path'" do
+      include_examples "adding files", "/path/to/my_file"
+    end
+
+    context "using a uri" do
+      include_examples "adding files", "ftp://localhost/foo/bar/baz"
+    end
+  end
+
+  describe "#download" do
+    let(:dest_path)   { Dir::Tmpname.create("") { |name| name } }
+    let(:source_file) { existing_ftp_file(10.megabytes) }
+    let(:source_path) { File.basename(source_file.path) }
+
+    after { File.delete(dest_path) if File.exist?(dest_path) }
+
+    it "downloads the file" do
+      subject.download(dest_path, source_path)
+
+      # Sanity check that what we are downloading is the size we expect
+      expect(source_path).to exist_on_ftp_server
+      expect(source_path).to have_size_on_ftp_server_of(10.megabytes)
+
+      expect(File.exist?(dest_path)).to be true
+      expect(File.stat(dest_path).size).to eq(10.megabytes)
+    end
+
+    it "can take input from a command" do
+      source_data = nil
+      subject.download(nil, source_path) do |input_writer|
+        source_data = `#{Gem.ruby} -e "print File.read('#{input_writer}')"`
+      end
+
+      # Nothing written, just printed the streamed file in the above command
+      expect(File.exist?(dest_path)).to be false
+      expect(source_path).to exist_on_ftp_server
```

Comment:
Shouldn't this be making some expectation on `source_data` (which I think is the data downloaded from the server) rather than `source_path` which definitely shouldn't have changed?

I may be missing some nuance here though.

Reply:
Sorry, this is a bit of a weird test.

I was mostly using these three tests as confirmation that the file existed in the first place, and that it was of the size we expected similar to the "sanity checks" in the above spec.

What I probably could do is label these three expects as `# sanity checks` and then have a final spec that takes `source_data`, and make sure it's `.size` is `10.megabytes`.  That make sense?

Unless I think of a more clever way of handling the `source_file` data (instead of it being `10.megabytes` of 0's), there isn't too much worth in validating it since it is so easy to generate.  Open to suggestions if we want to find a way to generate random data in a different, but some what speedy, way.  Considering some ideas ideas for that though, so I will report back if I do make a change.